flecsi/exec/hpx/task_prologue.hh

```diff
@@ -0,0 +1,223 @@
+/*
+    @@@@@@@@  @@           @@@@@@   @@@@@@@@ @@
+   /@@/////  /@@          @@////@@ @@////// /@@
+   /@@       /@@  @@@@@  @@    // /@@       /@@
+   /@@@@@@@  /@@ @@///@@/@@       /@@@@@@@@@/@@
+   /@@////   /@@/@@@@@@@/@@       ////////@@/@@
+   /@@       /@@/@@//// //@@    @@       /@@/@@
+   /@@       @@@//@@@@@@ //@@@@@@  @@@@@@@@ /@@
+   //       ///  //////   //////  ////////  //
+
+   Copyright (c) 2016, Triad National Security, LLC
+   All rights reserved.
+                                                                              */
+#pragma once
+
+/*! @file */
+
+#include <flecsi-config.h>
+
+#if !defined(FLECSI_ENABLE_HPX)
+#error FLECSI_ENABLE_HPX not defined! This file depends on HPX!
+#endif
+
+#include <hpx/modules/collectives.hpp>
+#include <hpx/modules/concurrency.hpp>
+#include <hpx/modules/serialization.hpp>
+
+#include "flecsi/data/privilege.hh"
+#include "flecsi/data/topology.hh"
+#include "flecsi/exec/hpx/future.hh"
+#include "flecsi/flog.hh"
+#include "flecsi/util/demangle.hh"
+
+#include <memory>
+#include <mutex>
+#include <string>
+
+namespace flecsi {
+namespace topo {
+struct global_base;
+}
+
+namespace exec {
+
+template<task_processor_type_t ProcessorType>
+struct task_prologue {
+
+private:
+  // the spinlock-pool is used to protect the initialization of the
+  // futures that are associated with the storage for a field
+  using spinlock_pool = ::hpx::util::spinlock_pool<::hpx::shared_future<void>>;
```

Comment:
This could go inside `visit` (where it wouldn't need an access-specifier and might not need a comment).